### PR TITLE
JAMES-3140 Remove SpamD stachTraces when running oneHundredMailsShoul…

### PR DESCRIPTION
**Pull request by chibenwa**
…dBeWellReceived

This test generates 100 starttraces.

Subpart of https://github.com/linagora/james-project/pull/3319

Why? Prevented me from getting valuable run log on the CI

### CODE REVIEW

**Comment by mbaechler** [comment]
> …dBeWellReceived
> 
> This test generates 100 starttraces.
> 
> Subpart of #3319
> 
> Why? Prevented me from getting valuable run log on the CI

It's disabling the listener not only for this test but for every other one of  `server/container/guice/cassandra-rabbitmq-guice`, right?

I don't know if it's something desirable but at least I would like this PR to express clearly it doesn't change the configuration for a single test

**Comment by chibenwa** [comment]
> It's disabling the listener not only for this test but for every other one of server/container/guice/cassandra-rabbitmq-guice, right?

Right.

A chance that we don't use it in any test ^^

> I don't know if it's something desirable but at least I would like this PR to express clearly it doesn't change the configuration for a single test

Let me reword the commit

**Comment by mbaechler** [comment]
> > It's disabling the listener not only for this test but for every other one of server/container/guice/cassandra-rabbitmq-guice, right?
> 
> Right.
> 
> A chance that we don't use it in any test ^^

Maybe we lack some test in our suite?

> > I don't know if it's something desirable but at least I would like this PR to express clearly it doesn't change the configuration for a single test
> 
> Let me reword the commit

We decided, in the first place, that this "plugin" was important enough to be included in the default setup we test with integration tests.

There are two assumptions here:
1. testing the product with plugins is relevant (it may detect bad things that would otherwise happen in production)
2. some plugins are actually so important that we may enable them by default in the product

For 1. I'm not sure: we are not mature enough to say that plugins won't break James but at the same time I can see it has the downside of make the tests slower and/or flood us with logs

For 2. it's an open discussion. I personally can't think to deploy a mail server without antispam and our SpamAssassin integration is nice. So I would push for inclusion.

WDYT?

**Comment by chibenwa** [comment]
> Maybe we lack some test in our suite?

Not really: it's tested in jmap integration tests.

> We decided, in the first place, that this "plugin" was important enough to be included in the default setup we test with integration tests.

I intend to challenge this once again.

I we really care about it, why do we run halth of it? If we really care, we'd start a SpamAssassin.

As we don't I conclude we don't care. 

Doing things halthway have a cost: these ugly stacktraces disturbing me debugging things.

1. Plugin cardinality is well... Overwhelming. nWe should test plugin in isolation to bring cardinality down. If we care then we should run tests in a @linagora @openpaas scope.

2. I might agree but clearly not for that one.

**Comment by mbaechler** [comment]
> > Maybe we lack some test in our suite?
> 
> Not really: it's tested in jmap integration tests.

Is it the right place?

> > We decided, in the first place, that this "plugin" was important enough to be included in the default setup we test with integration tests.
> 
> I intend to challenge this once again.
> 
> I we really care about it, why do we run halth of it? If we really care, we'd start a SpamAssassin.

I guess it's what we did with jmap integration tests. I don't know why we include the listener for this testsuite if it tests nothing.

> As we don't I conclude we don't care.

I would rather conclude we made a mistake, it's more likely

> Doing things halthway have a cost: these ugly stacktraces disturbing me debugging things.
> 
>     1. Plugin cardinality is well... Overwhelming. nWe should test plugin in isolation to bring cardinality down. If we care then we should run tests in a @linagora @OpenPaaS scope.

I don't get your argument, I never mentioned cardinality. Yes we have to tests plugins/mailet in isolation. But AFAIK, we have some enabled by default in our testsuite (quota, several mailets, etc)

The question, for me, is whether this SpamAssassin plugin should be enabled by default in the testsuite.

You should have noticed I said "I'm not sure" (:

>     2. I might agree but clearly not for that one.

That one what? Antispam?

**Comment by Arsnael** [comment]
Should SpamAssassin really be enabled by default? I'm not convinced to be honest. I will be honest to admit I don't know much about alternatives to SpamAssassin regarding antispam, but I would guess other softwares exist for that. 

I would guess if SpamAssassin is a plugin, it's because it lets the ability to people to develop plugins based on other antispams? Even if we don't have any others for now, it could still be a possibility right? Thus the plugin, thus can't be mandatory by definition?

**Comment by chibenwa** [comment]
> > > Maybe we lack some test in our suite?

>> Not really: it's tested in jmap integration tests.

> Is it the right place?

I can't tell. Probably not.

>> >We decided, in the first place, that this "plugin" was important enough to be included in the default setup we test with integration tests.

Well, weekend thoughts.

I do believe we actually did not had the tooling yet to enable/disable things as will.

>>  I intend to challenge this once again.
>> I we really care about it, why do we run halth of it? If we really care, we'd start a SpamAssassin.

> I guess it's what we did with jmap integration tests. I don't know why we include the listener for this testsuite if it tests nothing.

Well it's worse than it, we start spamassassin only for tests requiring it, the other tests have a listener that always fails.

>> As we don't I conclude we don't care.

> I would rather conclude we made a mistake, it's more likely

Both are not mutualy exclusive.

>>> Doing things halthway have a cost: these ugly stacktraces disturbing me debugging things.

>> 1. Plugin cardinality is well... Overwhelming. nWe should test plugin in isolation to bring cardinality down. If we care then we should run tests in a @linagora @OpenPaaS scope.

> I don't get your argument, I never mentioned cardinality. Yes we have to tests plugins/mailet in isolation. But AFAIK, we have some enabled by default in our testsuite (quota, several mailets, etc)
> The question, for me, is whether this SpamAssassin plugin should be enabled by default in the testsuite.
> You should have noticed I said "I'm not sure" (:

I'm sure we did not enabled it (ie start the corresponding docker everywhere) for runtime costs.

Then I believe we should assume the consequences of that choice: not add the corresponding listeners when we don't run spamassassin.

I *tired* of these noisy verbose stacktrazces I have any time I read a CI build, every time I debug a failed integration test.

Keeping these listeners hurts my productivity for zero benefits hence yes, I want to enact the fact that we *actually* never run them.

If that PR is rejected then shall we open a `add SpamAssassin container everywhere` PR ?

[...]

**Comment by nvduc91** [comment]
I dont really fully understand it but i believed if we treat `SpamAssassin` as a plugin so it should be `optional or not enabled by default`

**Comment by rouazana** [comment]
> Keeping these listeners hurts my productivity for zero benefits hence yes, I want to enact the fact that we actually never run them.

I don't agree with the "zero benefits", because I think having a full common pipeline has some value for our integration tests.
But as we test the full integration with SpamAssassin in an other place, and as it seems to do more harm (annoying stack traces) than good (have we only once get a bug catched by this full pipeline?) I agree to remove the listener in these tests.

**Comment by chibenwa** [comment]
Let's repĥrase it:

> have we only once get a bug catched by this full pipeline?

Have we only once get a bug catched by this full pipeline **with a disactivated feature related to that very feature**?

**Comment by chibenwa** [comment]
```
[5c452decd240c33200a5bb639dce022d9846bea8] [INFO] Running org.apache.james.backends.cassandra.TestingSessionTest
[5c452decd240c33200a5bb639dce022d9846bea8] 16:50:42.328 [ERROR] r.c.s.Schedulers - Scheduler worker in group main failed with an uncaught exception
[5c452decd240c33200a5bb639dce022d9846bea8] org.apache.james.backends.cassandra.Scenario$InjectedFailureException: Injected failure
[5c452decd240c33200a5bb639dce022d9846bea8] 	at org.apache.james.backends.cassandra.Scenario$Behavior.lambda$static$0(Scenario.java:45)
[5c452decd240c33200a5bb639dce022d9846bea8] 	at org.apache.james.backends.cassandra.Scenario$Behavior.lambda$awaitOn$1(Scenario.java:53)
[5c452decd240c33200a5bb639dce022d9846bea8] 	at org.apache.james.backends.cassandra.TestingSession.executeAsync(TestingSession.java:121)
[5c452decd240c33200a5bb639dce022d9846bea8] 	at org.apache.james.backends.cassandra.utils.CassandraAsyncExecutor.lambda$execute$0(CassandraAsyncExecutor.java:47)
[5c452decd240c33200a5bb639dce022d9846bea8] 	at reactor.core.publisher.MonoDefer.subscribe(MonoDefer.java:44)
[5c452decd240c33200a5bb639dce022d9846bea8] 	at reactor.core.publisher.Mono.subscribe(Mono.java:4210)
[5c452decd240c33200a5bb639dce022d9846bea8] 	at reactor.core.publisher.MonoIgnorePublisher.subscribe(MonoIgnorePublisher.java:47)
[5c452decd240c33200a5bb639dce022d9846bea8] 	at reactor.core.publisher.MonoCacheTime.subscribeOrReturn(MonoCacheTime.java:132)
[5c452decd240c33200a5bb639dce022d9846bea8] 	at reactor.core.publisher.Mono.subscribe(Mono.java:4195)
[5c452decd240c33200a5bb639dce022d9846bea8] 	at reactor.core.publisher.MonoSubscribeOn$SubscribeOnSubscriber.run(MonoSubscribeOn.java:124)
[5c452decd240c33200a5bb639dce022d9846bea8] 	at reactor.core.scheduler.WorkerTask.call(WorkerTask.java:84)
[5c452decd240c33200a5bb639dce022d9846bea8] 	at reactor.core.scheduler.WorkerTask.call(WorkerTask.java:37)
[5c452decd240c33200a5bb639dce022d9846bea8] 	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
[5c452decd240c33200a5bb639dce022d9846bea8] 	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
[5c452decd240c33200a5bb639dce022d9846bea8] 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
[5c452decd240c33200a5bb639dce022d9846bea8] 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
[5c452decd240c33200a5bb639dce022d9846bea8] 	at java.base/java.lang.Thread.run(Thread.java:834)
[5c452decd240c33200a5bb639dce022d9846bea8] Wrapped by: reactor.core.Exceptions$ErrorCallbackNotImplemented: org.apache.james.backends.cassandra.Scenario$InjectedFailureException: Injected failure
```

test this please

**Comment by chibenwa** [comment]
https://github.com/linagora/james-project/pull/3370 attends to fix the instability reported here

File: server/protocols/jmap-draft-integration-testing/jmap-draft-integration-testing-common/src/test/java/org/apache/james/jmap/draft/methods/integration/SpamAssassinModule.java [hunk 1]
```diff
@@ -49,6 +53,10 @@ protected void configure() {
                 new MailetConfigurationOverride(
                     org.apache.james.transport.mailets.SpamAssassin.class,
                     spamAssassinMailetConfig()));
+
+        bind(ListenersConfiguration.class)
```
Comment (mbaechler) [inline server/protocols/jmap-draft-integration-testing/jmap-draft-integration-testing-common/src/test/java/org/apache/james/jmap/draft/methods/integration/SpamAssassinModule.java:57]: I'm lost, why do you need that?

Comment (chibenwa) [reply server/protocols/jmap-draft-integration-testing/jmap-draft-integration-testing-common/src/test/java/org/apache/james/jmap/draft/methods/integration/SpamAssassinModule.java:57]: I can't inject the raw listener (as an extension it requires an executionMode that can't be injected as a core mailboxListener) 

Anyway now that we enforced it was a plugin ( cf https://github.com/linagora/james-project/pull/3359#pullrequestreview-409033388 ) the right testing strategy for plugin is to specify them in specific test via their POJO configuration, isn't it?

Comment (mbaechler) [reply server/protocols/jmap-draft-integration-testing/jmap-draft-integration-testing-common/src/test/java/org/apache/james/jmap/draft/methods/integration/SpamAssassinModule.java:57]: Can you put this piece of code into a commit with a proper explanation?

Explain why it's needed now, what is the problem, how you solved it.

Comment (chibenwa) [reply server/protocols/jmap-draft-integration-testing/jmap-draft-integration-testing-common/src/test/java/org/apache/james/jmap/draft/methods/integration/SpamAssassinModule.java:57]: Done with a wonderful 10 line git commit comment ;-)

Comment (mbaechler) [reply server/protocols/jmap-draft-integration-testing/jmap-draft-integration-testing-common/src/test/java/org/apache/james/jmap/draft/methods/integration/SpamAssassinModule.java:57]: the comment seems to be truncated

Comment (chibenwa) [reply server/protocols/jmap-draft-integration-testing/jmap-draft-integration-testing-common/src/test/java/org/apache/james/jmap/draft/methods/integration/SpamAssassinModule.java:57]: Crap

**Comment by chibenwa** [comment]
(no changes only a reword of the last commit.)

@mbaechler acceptable to you?

**Comment by chibenwa** [comment]
`MemorySpamAssassinContractTest.movingAMailToTrashShouldNotImpactSpamassassinLearning » ConditionTimeout`

test this please

**Comment by Arsnael** [comment]
Merged

**Pull request closed without merging**